pkg/bgp/speaker/speaker_test.go

```diff
@@ -0,0 +1,482 @@
+	// unblock our test. If the event never gets to our mock, a timeout
+	// will occur and fail the test.
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultTimeout)
+
+	var rr struct {
+		sync.Mutex
+		name string
+		svc  *metallbspr.Service
+		eps  *metallbspr.Endpoints
+	}
+
+	// when our mock metallb speaker has it's SetService method called this function will delegate for it,
+	// in the delegate we record the arguments with a response recorder for later retrieval and cancel the ctx
+	// effectively unblocking our test function.
+	mock := &mock.MockMetalLBSpeaker{
+		SetService_: func(name string, svc *metallbspr.Service, eps *metallbspr.Endpoints) types.SyncState {
+			rr.Lock()
+			rr.name, rr.svc, rr.eps = name, svc, eps
+			rr.Unlock()
+			// will unblock our test code.
+			cancel()
+			return types.SyncStateSuccess
+		},
+	}
+
+	// in this test we expect the service associated with the endpoints
+	// to exist.
+	//
+	// a service lookup is performed by OnUpdateEndpoint first.
+	spkr := &MetalLBSpeaker{
+		speaker:         mock,
+		announceLBIP:    true,
+		announcePodCIDR: true,
+		queue:           workqueue.New(),
+		services: map[k8s.ServiceID]*slim_corev1.Service{
+			serviceID: &service,
+		},
+	}
+
+	go spkr.run(ctx)
+
+	err := spkr.OnUpdateEndpoints(&slimendpoints)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	<-ctx.Done()
+	if ctx.Err() == context.DeadlineExceeded {
+		t.Fatal(errTimeout)
+	}
+
+	rr.Lock() // we'll do this just so race detector doen't bark at us.
+	defer rr.Unlock()
+
+	// confirm we recorded the correct metallb service name
+	// this should match our generated ServiceID.
+	if rr.name != serviceID.String() {
+		t.Fatalf("got: %v, want: %v", rr.name, serviceID.String())
+	}
+	if !cmp.Equal(rr.svc, &metallbService) {
+		t.Fatalf(cmp.Diff(rr.svc, metallbService))
+	}
+	if !cmp.Equal(rr.eps, &metallbendpoints) {
+		t.Fatalf(cmp.Diff(rr.eps, metallbendpoints))
+	}
+}
+
+// TestSpeakerOnUpdateNode confirms the speaker performs the correct
+// actions when an OnUpdateNode event takes place.
+//
+// This test effectively covers the OnAddNode as well, as this method
+// just deleagets to OnUpdateNode with nil as the first argument to
+// OnUpdateNode.
+func TestSpeakerOnUpdateNode(t *testing.T) {
+	// gen our test structures
+	node, advs := mock.GenTestNodeAndAdvertisements()
+
+	// our test will block on this ctx, our mock will cancel it and thus
+	// unblock our test. If the event never gets to our mock, a timeout
+	// will occur and fail the test.
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultTimeout)
+
+	var rr struct {
+		sync.Mutex
+		labels map[string]string
+		advs   []*metallbbgp.Advertisement
+	}
+
+	// launch a concurrent observer of callCount
+	// when the observer sees our mocks being called
+	// the necessary amount of times it'll cancel the ctx
+	// and unblock our test.
+	var callCount int32
+	go func() {
+		for atomic.LoadInt32(&callCount) != 3 {
```

Comment:
Why 3? It would be good to state that in the comment as well

Reply:
I see later down that we increment 3 times, but it wasn't obvious to me at first